pkg/cloud/aliyunprovider.go

```diff
@@ -0,0 +1,828 @@
+		ProviderID:       node.ProviderID,
+		RegionID:         node.RegionID,
+		InstanceType:     node.InstanceType,
+		OSType:           node.OSType,
+		OptimizedKeyword: optimizedKeyword,
+	}
+}
+
+func (aliyunNodeKey *AliyunNodeKey) ID() string {
+	return aliyunNodeKey.ProviderID
+}
+
+func (aliyunNodeKey *AliyunNodeKey) Features() string {
+	keyLookup := stringutil.DeleteEmptyStringsFromArray([]string{aliyunNodeKey.RegionID, aliyunNodeKey.InstanceType, aliyunNodeKey.OSType, aliyunNodeKey.OptimizedKeyword})
+	return strings.Join(keyLookup, "::")
+}
+
+func (aliyunNodeKey *AliyunNodeKey) GPUType() string {
+	return ""
+}
+
+func (aliyunNodeKey *AliyunNodeKey) GPUCount() int {
+	return 0
+}
+
+// Get's the key for the k8s node input
+func (aliyun *Aliyun) GetKey(mapValue map[string]string, node *v1.Node) Key {
+	//Mostly parse the Node object and get the ProviderID, region, InstanceType, OSType and OptimizedKeyword(In if block)
+	// Currently just hardcoding a Node but eventually need to Node object
+	slimK8sNode := generateSlimK8sNodeFromV1Node(node)
+
+	optimizedKeyword := ""
+	if slimK8sNode.IsIoOptimized {
+		optimizedKeyword = ALIYUN_OPTIMIZE_KEYWORD
+	} else {
+		optimizedKeyword = ALIYUN_NON_OPTIMIZE_KEYWORD
+	}
+	return NewAliyunNodeKey(slimK8sNode, optimizedKeyword)
+}
+
+type AliyunPVKey struct {
+	ProviderID       string
+	RegionID         string
+	DiskType         string
+	DiskCategory     string
+	PerformaceLevel  string
+	StorageClassName string
+}
+
+func (aliyunPVKey *AliyunPVKey) Features() string {
+	keyLookup := stringutil.DeleteEmptyStringsFromArray([]string{aliyunPVKey.RegionID, aliyunPVKey.DiskType, aliyunPVKey.DiskCategory, aliyunPVKey.PerformaceLevel})
+	return strings.Join(keyLookup, "::")
+}
+
+func (aliyunPVKey *AliyunPVKey) ID() string {
+	return aliyunPVKey.ProviderID
+}
+
+// Get storage class information for PV.
+func (aliyunPVKey *AliyunPVKey) GetStorageClass() string {
+	return aliyunPVKey.StorageClassName
+}
+
+// Helper functions for aliyunprovider.go
+
+// createDescribePriceACSRequest creates the HTTP GET request for the required resources' Price information,
+// When supporting subscription and Premptible resources this HTTP call needs to be modified with PriceUnit information
+// When supporting different new type of instances like Compute Optimized, Memory Optimized etc make sure you add the instance type
+// in unit test and check if it works or not to create the ack request and processDescribePriceAndCreateAliyunPricing function
+// else more paramters need to be pulled from kubernetes node response or gather infromation from elsewhere and function modified.
+// TO-DO: Add disk adjustments to the node , Test it out!
+func createDescribePriceACSRequest(i interface{}) (*requests.CommonRequest, error) {
+	request := requests.NewCommonRequest()
+	request.Method = requests.GET
+	request.Product = ALIYUN_ECS_PRODUCT_CODE
+	request.Domain = ALIYUN_ECS_DOMAIN
+	request.Version = ALIYUN_ECS_VERSION
```

Comment:
Is it "ECS" or "ACS"? Or are they separate things?

Reply:
ECS is the Elastic compute service like EC2, and this PRODUCT_CODE which is "ecs" is used to get the price of the backing instance type of the node. I tried searching for "ACS" product on alibaba cloud and its not present.

Reply:
Okay, should we call this function `createDescribePriceECSRequest`, then? (rather than `...ACSRequest`)

Reply:
Ah now i see where you referred ACS, https://github.com/aliyun/alibaba-cloud-sdk-go/blob/v1.62.26/sdk/requests/acs_request.go#L61, this is the interface for Alibaba Cloud Services (ACS) Request and Describe Price is one of such request Action name or API name on Product code ECS. So maybe the name seems fine but im okay to change it.

Reply:
Aha! All becomes clear, thank you